packages/providers/src/transaction-request/transaction-request.ts

```diff
@@ -226,130 +228,188 @@ export abstract class BaseTransactionRequest implements BaseTransactionRequestLi
+    } else {
+      witnessIndex = this.getCoinInputWitnessIndexByOwner(recipient);
+
+      // Insert a dummy witness if no witness exists
+      if (typeof witnessIndex !== 'number') {
+        witnessIndex = this.createWitness();
+      }
     }
 
+    const input: MessageTransactionRequestInput = {
+      ...message,
+      type: InputType.Message,
+      sender: sender.toB256(),
+      recipient: recipient.toB256(),
+      amount,
+      witnessIndex,
+      predicate,
+      predicateData,
+    };
+
     // Insert the Input
-    this.pushInput(
-      isCoin(resource)
-        ? ({
-            type,
-            ...resource,
-            owner: resource.owner.toB256(),
-            witnessIndex,
-            txPointer: '0x00000000000000000000000000000000',
-          } as CoinTransactionRequestInput)
-        : ({
-            type,
-            ...resource,
-            sender: resource.sender.toB256(),
-            recipient: resource.recipient.toB256(),
-            witnessIndex,
-            txPointer: '0x00000000000000000000000000000000',
-          } as MessageTransactionRequestInput)
-    );
+    this.pushInput(input);
 
-    // Find the ChangeOutput for the AssetId of the Resource
-    const changeOutput = this.getChangeOutputs().find(
-      (output) => hexlify(output.assetId) === assetId
-    );
+    // Insert a ChangeOutput if it does not exist
+    this.addChangeOutput(recipient, assetId);
+  }
 
-    // Throw if the existing ChangeOutput is not for the same owner
-    if (changeOutput && hexlify(changeOutput.to) !== ownerAddress.toB256()) {
-      throw new ChangeOutputCollisionError();
+  /**
+   * Adds a single resource to the transaction by adding a coin/message input and a
+   * change output for the related assetId, if one it was not added yet.
+   *
+   * @param resource - The resource to add.
+   * @returns This transaction.
+   */
+  addResource(resource: Resource) {
+    if (isCoin(resource)) {
+      this.addCoinInput(resource);
+    } else {
+      this.addMessageInput(resource);
     }
 
-    // Insert a ChangeOutput if it does not exist
-    if (!changeOutput) {
-      this.pushOutput({
-        type: OutputType.Change,
-        to: ownerAddress.toB256(),
-        assetId,
-      });
+    return this;
+  }
+
+  /**
+   * Adds multiple resources to the transaction by adding coin/message inputs and change
+   * outputs from the related assetIds.
+   *
+   * @param resources - The resources to add.
+   * @returns This transaction.
+   */
+  addResources(resources: ReadonlyArray<Resource>) {
+    resources.forEach((resource) => this.addResource(resource));
+
+    return this;
+  }
+
+  /**
+   * Adds multiple resources to the transaction by adding coin/message inputs and change
+   * outputs from the related assetIds.
+   *
+   * @param resources - The resources to add.
+   * @returns This transaction.
+   */
+  addPredicateResource(resource: Resource, predicate: BytesLike, predicateData?: BytesLike) {
```

Comment:
If we have a predicate that requires data, the user still wouldn't have to provide the data because it's optional, and the transaction would fail.

Why not do:
```ts
addPredicateResources(resources, predicate: Predicate)
```

and then you'll have both `predicate.bytes` and `predicate.predicateData` available and this would no longer be an issue?

Reply:
@nedsalk I had the same idea. 

However, the package `predicate` already makes use of the package `provider`, where the `BaseTransactionRequest` lives. So I ended up receiving the following:

![Screenshot from 2023-08-30 10-26-18](https://github.com/FuelLabs/fuels-ts/assets/30977845/f6d17f68-e10e-41ea-a156-1a74eef7d9e4)

Reply:
It's weird that the `BaseTransactionRequest` says, "you can give me a predicate" via `addPredicateResource` but it itself doesn't know what a `Predicate` is (because it's got no dependency to the `predicate` package). It seems to me that the root issue is the `interfaces` package which is obfuscating all of these dependencies and we don't know what is **really** dependent on what so that we can more easily split packages, move code from one package to another, etc.

Now that I think of it, if we had the `Predicate` available, we could have used only the `addResources` method, made it generic and created a special input type for this method which checks if the source of the resources is a predicate and, if it is, mandate the passing of that predicate as well.

Reply:
It's weird that the `BaseTransactionRequest` says, "you can give me a predicate" via `addPredicateResource` but it itself doesn't know what a predicate is (because it's got no dependency to the `predicate` package). It seems to me that the root issue is the `interfaces` package which is obfuscating all of these dependencies and we don't know what is really dependent on what so that we can more easily split packages, move code from one package to another, etc.